controllers/storageautoscaler/scraper.go

```diff
@@ -0,0 +1,151 @@
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdUsageQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return
+	}
+	// refresh the sync map
+	r.SyncMap = &sync.Map{}
+	// update the sync map with the highest value of osd usage per device class
+	updateSyncMap(metrics, r.SyncMap)
+	// get the storage autoscaler objects
+	objectList, err := getStorageAutoScalerObjects(ctx, r.Client)
+	if err != nil {
+		r.Log.Error(err, "failed to get resource objects")
+		return
+	}
+	// filter the objects for scaling
+	filteredObjectList := filterObjectsForScaling(objectList, r.SyncMap, r.Log)
+
+	// send generic events for the filtered objects
+	sendGenericEvent(filteredObjectList, r.EventCh, r.Log)
+}
+
+// update the sync map with the highest value of osd usage per device class
+func updateSyncMap(metrics model.Vector, syncMap *sync.Map) {
+	// create a map of device class to highest osd usage
+	deviceClassUsage := make(map[string]float64)
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		if _, ok := deviceClassUsage[string(deviceClass)]; ok {
+			if float64(osd.Value) > deviceClassUsage[string(deviceClass)] {
+				deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+			}
+		} else {
+			deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+		}
+	}
+
+	// update the sync map
+	for deviceClass, usage := range deviceClassUsage {
+		syncMap.Store(deviceClass, usage)
+	}
+}
+
+func getStorageAutoScalerObjects(ctx context.Context, client client.Client) (*ocsv1.StorageAutoScalerList, error) {
+	// get all the storage autoscaler cr objects
+	storageAutoScalerList := &ocsv1.StorageAutoScalerList{}
+	err := client.List(ctx, storageAutoScalerList)
+	if err != nil {
+		return nil, err
+	}
+
+	return storageAutoScalerList, nil
+}
+
+func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList, syncMap *sync.Map, log logr.Logger) []struct {
+	namespace string
+	name      string
+} {
+	var objectList []struct {
+		namespace string
+		name      string
+	}
+
+	for _, storageAutoScaler := range storageAutoScalerList.Items {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		deviceClassThreshold := storageAutoScaler.Spec.StorageScalingThresholdPercent
+		// get the osd usage from the sync map
+		usage, ok := syncMap.Load(deviceClass)
+		if !ok {
+			log.Error(fmt.Errorf("osd usage not found for device class"), "osd usage not found for device class", "device class", deviceClass)
+			continue
+		}
+
+		if usage == nil {
+			err := fmt.Errorf("osd usage is nil for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
+			log.Error(err, "device class not found in sync map")
+			continue
+		}
+
+		deviceClassUsage := usage.(float64)
+		if (deviceClassUsage * 100) >= float64(deviceClassThreshold) {
+			objectList = append(objectList, struct {
```

Comment:
one more time, creating anonymous struct, even if `types.NamespacedName` isn't being used, having this as a local type would've been beneficial.